sklearn/metrics/classification.py

```diff
@@ -391,72 +391,269 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+
+    if average == 'binary':
+        if y_type == 'binary':
+            if pos_label not in present_labels:
+                if len(present_labels) < 2:
+                    return 0.
+                else:
+                    raise ValueError("pos_label=%r is not a valid label: "
+                                    "%r" % (pos_label, present_labels))
+            labels = [pos_label]
+        else:
+            raise ValueError("Target is %s but average='binary'. Please "
+                            "choose another average setting." % y_type)
+    elif pos_label not in (None, 1):
+        warnings.warn("Note that pos_label (set to %r) is ignored when "
+                      "average != 'binary' (got %r). You may use "
+                      "labels=[pos_label] to specify a single positive class."
+                      % (pos_label, average), UserWarning)
+
+    if labels is None:
+        labels = present_labels
+        n_labels = None
+    else:
+        n_labels = len(labels)
+        labels = np.hstack([labels, np.setdiff1d(present_labels, labels,
+                                                 assume_unique=True)])
+
     if y_type.startswith('multilabel'):
+        if average in (None, 'samples'):
+            if normalize is None:
+                # default is average='samples'
+                average = 'samples'
+            else:
+                if normalize:
+                    average = 'samples'
+                else:
+                    average = 'none-samples'
+                warn_message = ("'normalize' was removed in version 0.20 and "
+                                "will be removed in 0.22, instead use "
+                                "`average='%s'`." % normalize)
+                warnings.warn(warn_message, DeprecationWarning)
+        # else:
+            # otherwise what should we do? raise warning or ValueError?
+
+        if not np.all(labels == present_labels):
+            if np.max(labels) > np.max(present_labels):
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d > %d' %
+                                 (np.max(labels), np.max(present_labels)))
+            if np.min(labels) < 0:
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d < 0' % np.min(labels))
+
+        # wait for response on 'prf-bug' PR, since I'm less than 90% sure
+        if n_labels is not None:
+            y_true = y_true[:, labels[:n_labels]]
+            y_pred = y_pred[:, labels[:n_labels]]
+
         with np.errstate(divide='ignore', invalid='ignore'):
-            # oddly, we may get an "invalid" rather than a "divide" error here
-            pred_or_true = count_nonzero(y_true + y_pred, axis=1)
-            pred_and_true = count_nonzero(y_true.multiply(y_pred), axis=1)
-            score = pred_and_true / pred_or_true
-            score[pred_or_true == 0.0] = 1.0
+
```

Comment:
You basically have three cases: none, macro and micro.
For macro, you default to class_weight=None and axis=0, but change these in the 'weighted' and 'samples' cases.
For None you similarly switch the 'none-samples' axis.

Reply:
Done this. As well as related comments to *simplify* the multilabel code of using `count_nonzero`.